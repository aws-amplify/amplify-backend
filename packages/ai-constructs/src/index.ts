import * as conversation from './conversation';

export { conversation };
