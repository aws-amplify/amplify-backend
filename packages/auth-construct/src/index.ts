export {
  AuthProps,
  EmailLogin,
  EmailLoginSettings,
  GoogleProviderProps,
  FacebookProviderProps,
  AmazonProviderProps,
  AppleProviderProps,
  OidcProviderProps,
  SamlProviderProps,
  ExternalProviderOptions,
  VerificationEmailWithCode,
  VerificationEmailWithLink,
  MFA,
  MFASettings,
  PhoneNumberLogin,
  TriggerEvent,
  IdentityProviderProps,
  AttributeMapping,
} from './types.js';
export { AmplifyAuth } from './construct.js';
export { triggerEvents } from './trigger_events.js';
