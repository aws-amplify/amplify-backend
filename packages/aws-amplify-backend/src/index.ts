export * from '@aws-amplify/backend';
