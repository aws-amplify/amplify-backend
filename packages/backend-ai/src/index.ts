export * from './types.js';

export { getMessage } from './get_message.js';
