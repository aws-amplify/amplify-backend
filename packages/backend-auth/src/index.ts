export { BackendAuth, AmplifyAuthProps, defineAuth } from './factory.js';
export * from './types.js';
