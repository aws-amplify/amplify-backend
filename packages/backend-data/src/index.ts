export { defineData } from './factory.js';
export {
  ApiKeyAuthorizationModeProps,
  LambdaAuthorizationModeProps,
  OIDCAuthorizationModeProps,
  DefaultAuthorizationMode,
  AuthorizationModes,
  DataSchemaInput,
  DataProps,
} from './types.js';
