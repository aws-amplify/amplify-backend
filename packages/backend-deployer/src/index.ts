export * from './cdk_deployer_singleton_factory.js';
export * from './types.js';
