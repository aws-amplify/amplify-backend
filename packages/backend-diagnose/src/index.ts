export { getCdkInfo, formatCdkInfo } from './cdk_info.js';
export { getEnvInfo, formatEnvInfo } from './env_info.js';
