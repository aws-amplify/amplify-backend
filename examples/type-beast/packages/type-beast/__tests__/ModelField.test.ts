import { expectTypeTestsToPassAsync } from 'jest-tsd';
import a from '../index';
import { ModelField, InternalField, __auth } from '../src/ModelField';
import {
  ModelRelationalField,
  InternalRelationalField,
} from '../src/ModelRelationalField';
import { Authorization, ImpliedAuthFields, __data } from '../src/Authorization';

// evaluates type defs in corresponding test-d.ts file
it('should not produce static type errors', async () => {
  await expectTypeTestsToPassAsync(__filename);
});

describe('field level auth', () => {
  it('implied field types can be inferred', () => {
    const field = a
      .string()
      .array()
      .authorization([
        a.allow.public().to(['read']),
        a.allow.private().to(['read', 'create']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete']),
      ]);

    type ExpectedAuthFields = typeof field extends ModelField<
      any,
      any,
      infer Auth
    >
      ? Auth extends Authorization<any, any>
        ? ImpliedAuthFields<Auth>
        : never
      : never;

    const authFields: ExpectedAuthFields = {
      owner: 'a string',
      admin: ['array', 'of', 'string'],
    };

    const brokenAuthFieldsA: ExpectedAuthFields = {
      // @ts-expect-error
      owner: ['array', 'of', 'string'],
    };

    const brokenAuthFieldsB: ExpectedAuthFields = {
      // @ts-expect-error
      admin: 'just a regular string',
    };
  });

  it('implied field types can be inferred from related model fields', () => {
    const field = a
      .belongsTo('Widget')
      .authorization([
        a.allow.public().to(['read']),
        a.allow.private().to(['read', 'create']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete']),
      ]);

    type ExpectedAuthFields = typeof field extends ModelRelationalField<
      any,
      any,
      any,
      infer Auth
    >
      ? Auth extends Authorization<any, any>
        ? ImpliedAuthFields<Auth>
        : never
      : never;

    const authFields: ExpectedAuthFields = {
      owner: 'a string',
      admin: ['array', 'of', 'string'],
    };

    const brokenAuthFieldsA: ExpectedAuthFields = {
      // @ts-expect-error
      owner: ['array', 'of', 'string'],
    };

    const brokenAuthFieldsB: ExpectedAuthFields = {
      // @ts-expect-error
      admin: 'just a regular string',
    };
  });

  it('implied fields objects can be extracted', () => {
    const field = a
      .string()
      .array()
      .authorization([
        a.allow.public().to(['read']),
        a.allow.private().to(['read', 'create']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete'])
          .identityClaim('identityClaimValue'),
      ]) as InternalField;

    expect(field.data.authorization).toMatchSnapshot();
  });

  it('implied fields objects can be extracted from related model fields', () => {
    const field = a
      .belongsTo('Widget')
      .authorization([
        a.allow.public().to(['read']),
        a.allow.private().to(['read', 'create']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete']),
        a.allow
          .multipleOwners()
          .inField('admin')
          .to(['read', 'create', 'delete'])
          .identityClaim('identityClaimValue'),
      ]) as InternalRelationalField;

    expect(field.data.authorization).toMatchSnapshot();
  });
});
