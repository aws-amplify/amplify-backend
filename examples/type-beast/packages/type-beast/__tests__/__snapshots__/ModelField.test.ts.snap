// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field level auth implied fields objects can be extracted 1`] = `
[
  {
    Symbol(data): {
      "groupOrOwnerField": undefined,
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": false,
      "operations": [
        "read",
      ],
      "provider": undefined,
      "strategy": "public",
    },
  },
  {
    Symbol(data): {
      "groupOrOwnerField": undefined,
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": false,
      "operations": [
        "read",
        "create",
      ],
      "provider": undefined,
      "strategy": "private",
    },
  },
  {
    "identityClaim": [Function],
    Symbol(data): {
      "groupOrOwnerField": "admin",
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": true,
      "operations": [
        "read",
        "create",
        "delete",
      ],
      "provider": undefined,
      "strategy": "owner",
    },
  },
  {
    Symbol(data): {
      "groupOrOwnerField": "admin",
      "groups": undefined,
      "identityClaim": "identityClaimValue",
      "multiOwner": true,
      "operations": [
        "read",
        "create",
        "delete",
      ],
      "provider": undefined,
      "strategy": "owner",
    },
  },
]
`;

exports[`field level auth implied fields objects can be extracted from related model fields 1`] = `
[
  {
    Symbol(data): {
      "groupOrOwnerField": undefined,
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": false,
      "operations": [
        "read",
      ],
      "provider": undefined,
      "strategy": "public",
    },
  },
  {
    Symbol(data): {
      "groupOrOwnerField": undefined,
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": false,
      "operations": [
        "read",
        "create",
      ],
      "provider": undefined,
      "strategy": "private",
    },
  },
  {
    "identityClaim": [Function],
    Symbol(data): {
      "groupOrOwnerField": "admin",
      "groups": undefined,
      "identityClaim": undefined,
      "multiOwner": true,
      "operations": [
        "read",
        "create",
        "delete",
      ],
      "provider": undefined,
      "strategy": "owner",
    },
  },
  {
    Symbol(data): {
      "groupOrOwnerField": "admin",
      "groups": undefined,
      "identityClaim": "identityClaimValue",
      "multiOwner": true,
      "operations": [
        "read",
        "create",
        "delete",
      ],
      "provider": undefined,
      "strategy": "owner",
    },
  },
]
`;
